# Task 2

# Car Price Prediction with Machine Learning

# Load Data

In [121]:
import pandas as pd
data = pd.read_csv(r"C:\Users\hp\Downloads\car data.csv")
data.head(3)

Car_Name  Year  Selling_Price  Present_Price  Driven_kms Fuel_Type  \
0     ritz  2014           3.35           5.59       27000    Petrol   
1      sx4  2013           4.75           9.54       43000    Diesel   
2     ciaz  2017           7.25           9.85        6900    Petrol   

  Selling_type Transmission  Owner  
0       Dealer       Manual      0  
1       Dealer       Manual      0  
2       Dealer       Manual      0

# Preprocessing in data

# Check null values

In [122]:
data.isnull().sum()

Car_Name         0
Year             0
Selling_Price    0
Present_Price    0
Driven_kms       0
Fuel_Type        0
Selling_type     0
Transmission     0
Owner            0
dtype: int64

In [123]:
data.describe()

Year  Selling_Price  Present_Price     Driven_kms       Owner
count   301.000000     301.000000     301.000000     301.000000  301.000000
mean   2013.627907       4.661296       7.628472   36947.205980    0.043189
std       2.891554       5.082812       8.642584   38886.883882    0.247915
min    2003.000000       0.100000       0.320000     500.000000    0.000000
25%    2012.000000       0.900000       1.200000   15000.000000    0.000000
50%    2014.000000       3.600000       6.400000   32000.000000    0.000000
75%    2016.000000       6.000000       9.900000   48767.000000    0.000000
max    2018.000000      35.000000      92.600000  500000.000000    3.000000

In [124]:
data.columns

Index(['Car_Name', 'Year', 'Selling_Price', 'Present_Price', 'Driven_kms',
       'Fuel_Type', 'Selling_type', 'Transmission', 'Owner'],
      dtype='object')

In [125]:
data.shape

(301, 9)

In [126]:
data.dtypes

Car_Name          object
Year               int64
Selling_Price    float64
Present_Price    float64
Driven_kms         int64
Fuel_Type         object
Selling_type      object
Transmission      object
Owner              int64
dtype: object

In [127]:
# drop not meaningful column

In [128]:
data.drop(columns = ['Car_Name'], inplace = True)

In [129]:
data.head(3)

Year  Selling_Price  Present_Price  Driven_kms Fuel_Type Selling_type  \
0  2014           3.35           5.59       27000    Petrol       Dealer   
1  2013           4.75           9.54       43000    Diesel       Dealer   
2  2017           7.25           9.85        6900    Petrol       Dealer   

  Transmission  Owner  
0       Manual      0  
1       Manual      0  
2       Manual      0

# Convert the categorical data into numarical data 

In [130]:

le_fuel = LabelEncoder()
data['Fuel_Type'] = le_fuel.fit_transform(data['Fuel_Type'])

le_sell = LabelEncoder()
data['Selling_type'] = le_sell.fit_transform(data['Selling_type'])

le_trans = LabelEncoder()
data['Transmission'] = le_trans.fit_transform(data['Transmission'])


In [131]:
#from sklearn.preprocessing import LabelEncoder

#for i in ['Fuel_Type','Selling_type','Transmission']:
 #   le = LabelEncoder()
  #  data[i]= le.fit_transform(data[i])

In [132]:
data.head(3)

Year  Selling_Price  Present_Price  Driven_kms  Fuel_Type  Selling_type  \
0  2014           3.35           5.59       27000          2             0   
1  2013           4.75           9.54       43000          1             0   
2  2017           7.25           9.85        6900          2             0   

   Transmission  Owner  
0             1      0  
1             1      0  
2             1      0

# Convert the data into x and y

In [133]:
x = data.drop(columns= ['Selling_Price'])
y = data['Selling_Price']

# Scaling 

In [134]:
from sklearn.preprocessing import StandardScaler

In [135]:
ss = StandardScaler()
x = pd.DataFrame(ss.fit_transform(x))
x

0         1         2         3         4        5         6
0    0.128897 -0.236256 -0.256224  0.500183 -0.737285  0.39148 -0.174501
1   -0.217514  0.221544  0.155911 -1.852241 -0.737285  0.39148 -0.174501
2    1.168129  0.257472 -0.773969  0.500183 -0.737285  0.39148 -0.174501
3   -0.910335 -0.403151 -0.817758  0.500183 -0.737285  0.39148 -0.174501
4    0.128897 -0.087906  0.141743 -1.852241 -0.737285  0.39148 -0.174501
..        ...       ...       ...       ...       ...      ...       ...
296  0.821718  0.460295 -0.076225 -1.852241 -0.737285  0.39148 -0.174501
297  0.475308 -0.200328  0.593804  0.500183 -0.737285  0.39148 -0.174501
298 -1.603156  0.390756  1.313340  0.500183 -0.737285  0.39148 -0.174501
299  1.168129  0.564604 -0.719876 -1.852241 -0.737285  0.39148 -0.174501
300  0.821718 -0.200328 -0.810958  0.500183 -0.737285  0.39148 -0.174501

[301 rows x 7 columns]

# Split the data for training and testing

In [136]:
from sklearn.model_selection import train_test_split

In [137]:
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.2,random_state=42)

# Apply ML algorithm for train the model 

In [138]:
from sklearn.ensemble import RandomForestRegressor
model1 = RandomForestRegressor()
model1.fit(x_train, y_train)

RandomForestRegressor()

In [139]:
print("Testing accuracy is:",model1.score(x_test,y_test)*100)
print("Training accuracy is:",model1.score(x_train,y_train)*100)

Testing accuracy is: 96.17518486309521
Training accuracy is: 98.33972300602792


In [140]:
y_pred = model1.predict(x_test)
y_pred

array([ 0.4218, 11.1122,  4.7615,  0.209 ,  7.4916,  6.2965,  1.112 ,
        0.5712,  0.4677,  6.7345,  7.9033,  1.071 ,  7.8981,  0.4508,
        5.2685,  2.508 ,  1.1781, 14.9934,  0.4692,  1.5835,  0.353 ,
        8.1079,  4.649 ,  2.7735,  0.5006,  3.5075,  5.275 ,  3.147 ,
        1.2229,  1.1701,  0.4198,  9.5945,  0.448 ,  2.5915,  7.7664,
        4.2815,  6.24  ,  6.6744,  2.5085,  7.0239,  4.1966,  3.4146,
        4.931 ,  0.5615,  6.1935,  0.7276,  8.0805,  7.677 ,  2.84  ,
        3.515 ,  4.9335,  1.524 , 22.3423, 21.1113,  6.328 ,  9.9609,
        5.1095,  8.9147,  2.653 ,  6.8555,  0.2285])

In [141]:
y_test

177     0.35
289    10.11
228     4.95
198     0.15
60      6.95
       ...  
234     5.50
296     9.50
281     2.10
285     7.40
182     0.30
Name: Selling_Price, Length: 61, dtype: float64

# Check the error in model

In [142]:
from sklearn.metrics import mean_absolute_error,mean_squared_error

In [143]:
mean_absolute_error(y_test,model1.predict(x_test))

0.6304000000000003

In [144]:
mean_squared_error(y_test, model1.predict(x_test))

0.8810685236065581

# Calculate R2 score 

In [145]:
from sklearn.metrics import r2_score

train_preds = model1.predict(x_train)
r2 = r2_score(y_train, train_preds)
print("Training R² Score:", r2)  # Should be close to 1

Training R² Score: 0.9833972300602792


# Predict the output 

In [146]:
# we predict here after scaling data input
# 0.128897	-0.236256	-0.256224	0.500183	-0.737285	0.39148	-0.174501 and our output is 3.35 

In [147]:
model1.predict([[0.128897,	-0.236256,	-0.256224,	0.500183,	-0.737285,	0.39148, -0.174501]])

array([3.83])

In [148]:
model1.predict([[0.821718,0.692093,-0.668875,0.500183,-0.737285,0.391480,-0.174501]])

array([11.1122])

## for example we take user data to predict the output 

In [149]:
#	2014,	5.59	27000	Petrol	Dealer	Manual	0 its selling price is 3.35

In [150]:

feature_names = ['Year', 'Present_Price', 'Driven_kms', 
                 'Fuel_Type', 'Selling_type', 'Transmission', 'Owner']


In [151]:
user_data = pd.DataFrame([[2014,5.59,27000,'Petrol','Dealer','Manual',0]],columns = feature_names)
user_data  

Year  Present_Price  Driven_kms Fuel_Type Selling_type Transmission  Owner
0  2014           5.59       27000    Petrol       Dealer       Manual      0

# Encode the user data 

In [152]:

# Encode correctly
user_data['Fuel_Type'] = le_fuel.transform(user_data['Fuel_Type'])
user_data['Selling_type'] = le_sell.transform(user_data['Selling_type'])
user_data['Transmission'] = le_trans.transform(user_data['Transmission'])

In [153]:

#for i in ['Fuel_Type','Selling_type','Transmission']:
 #   user_data[i]= le.transform(user_data[i])

#user_data

# Scaling 

In [154]:

user_data_scaled = ss.transform(user_data)
user_data_scaled


array([[ 0.128897  , -0.23625646, -0.25622446,  0.50018321, -0.73728539,
         0.39148015, -0.17450057]])

In [155]:
# prediction
model1.predict(user_data_scaled)

array([3.83])

## our user data output has come 3.83 and our original output was 3.35 so beasically it's close and good to predict the output 

# Thank You!